neonvm/controllers/vm_controller_unit_test.go

```diff
@@ -0,0 +1,262 @@
+	}
+
+	return params
+}
+
+func (p *testParams) initVM(vm *vmv1.VirtualMachine) *vmv1.VirtualMachine {
+	err := p.client.Create(p.ctx, vm)
+	require.NoError(p.t, err)
+	p.origVM = vm
+
+	// Do serialize/deserialize, to normalize resource.Quantity
+	return p.getVM()
+}
+
+func (p *testParams) getVM() *vmv1.VirtualMachine {
+	var obj vmv1.VirtualMachine
+	err := p.client.Get(p.ctx, client.ObjectKeyFromObject(p.origVM), &obj)
+	require.NoError(p.t, err)
+
+	return &obj
+}
+
+func TestReconcile(t *testing.T) {
+	params := newTestParams(t)
+	origVM := params.initVM(defaultVm())
+
+	req := reconcile.Request{
+		NamespacedName: client.ObjectKeyFromObject(origVM),
+	}
+
+	// Round 1
+	res, err := params.r.Reconcile(params.ctx, req)
+	assert.NoError(t, err)
+
+	// Added finalizer
+	assert.Equal(t, reconcile.Result{
+		Requeue: true,
+	}, res)
+	assert.Contains(t, params.getVM().Finalizers, virtualmachineFinalizer)
+
+	// Round 2
+	res, err = params.r.Reconcile(params.ctx, req)
+	assert.NoError(t, err)
+	assert.Equal(t, false, res.Requeue)
+
+	// VM is pending
+	assert.Equal(t, vmv1.VmPending, params.getVM().Status.Phase)
+
+	// Round 3
+	params.mockRecorder.On("Event", mock.Anything, "Normal", "Created",
+		mock.Anything)
+	res, err = params.r.Reconcile(params.ctx, req)
+	assert.NoError(t, err)
+	assert.Equal(t, false, res.Requeue)
+
+	// We now have a pod
+	vm := params.getVM()
+	assert.NotEmpty(t, vm.Status.PodName)
+	// Spec is unchanged
+	assert.Equal(t, vm.Spec, origVM.Spec)
+
+	// Round 4
+	res, err = params.r.Reconcile(params.ctx, req)
+	assert.NoError(t, err)
+	assert.Equal(t, false, res.Requeue)
+
+	// Pod is pending, so nothing changes
+	assert.Equal(t, vm, params.getVM())
+}
+
+func PrettyPrint(t *testing.T, obj any) {
+	s, _ := json.MarshalIndent(obj, "", "\t")
+	t.Logf("%s\n", s)
+}
+
+func TestRunningPod(t *testing.T) {
+	params := newTestParams(t)
+	origVM := defaultVm()
+	origVM.Finalizers = append(origVM.Finalizers, virtualmachineFinalizer)
+	origVM.Status.Phase = vmv1.VmPending
+
+	origVM = params.initVM(origVM)
+
+	req := reconcile.Request{
+		NamespacedName: client.ObjectKeyFromObject(origVM),
+	}
+
+	// Round 1
+	params.mockRecorder.On("Event", mock.Anything, "Normal", "Created",
+		mock.Anything)
+	res, err := params.r.Reconcile(params.ctx, req)
+	require.NoError(t, err)
+	assert.Equal(t, false, res.Requeue)
+
+	// We now have a pod
+	podName := params.getVM().Status.PodName
+	podKey := client.ObjectKey{
+		Namespace: origVM.Namespace,
+		Name:      podName,
+	}
+	var pod corev1.Pod
+	err = params.client.Get(params.ctx, podKey, &pod)
+	require.NoError(t, err)
+
+	assert.Len(t, pod.Spec.Containers, 1)
+	assert.Equal(t, "neonvm-runner", pod.Spec.Containers[0].Name)
+	assert.Equal(t, "vm-runner-img", pod.Spec.Containers[0].Image)
+	assert.Len(t, pod.Spec.InitContainers, 2)
+	assert.Equal(t, "init", pod.Spec.InitContainers[0].Name)
+	assert.Equal(t, "init-kernel", pod.Spec.InitContainers[1].Name)
+
+	PrettyPrint(t, pod)
```

Comment:
This is just for debugging, right? is it intentionally left in? (I could go either way, just curious what you think)

Reply:
Ack, fair enough